app/drivers/zephyr/battery_voltage_divider.c

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_battery_voltage_divider
+
+#include <device.h>
+#include <drivers/kscan.h>
+#include <drivers/gpio.h>
+#include <drivers/adc.h>
+#include <drivers/sensor.h>
+#include <logging/log.h>
+#include <math.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+struct io_channel_config {
+    const char *label;
+    uint8_t channel;
+};
+
+struct gpio_channel_config {
+    const char *label;
+    uint8_t pin;
+    uint8_t flags;
+};
+
+struct bvd_config {
+    struct io_channel_config io_channel;
+    struct gpio_channel_config power_gpios;
+    uint32_t output_ohm;
+    uint32_t full_ohm;
+};
+
+struct bvd_data {
+    struct device *adc;
+    struct device *gpio;
+    struct adc_channel_cfg acc;
+    struct adc_sequence as;
+    uint16_t adc_raw;
+    uint16_t voltage;
+    uint8_t state_of_charge;
+};
+
+static uint8_t lithium_ion_mv_to_pct(int16_t bat_mv) {
+    // Magic function that maps mV to this discharge graph from adafruit:
+    // https://learn.adafruit.com/li-ion-and-lipoly-batteries/voltages
+    return round(106.818 +
+                 (-0.032685 - 106.818) / pow(1 + pow(bat_mv / 3679.35, 58.979), 0.347386));
+}
+
+static int bvd_sample_fetch(struct device *dev, enum sensor_channel chan) {
```

Comment:
You should be checking the specific requested channel, and return `-ENOTSUP` if not the battery channel.